### PR TITLE
Timx 425 update load dataset and apply filtering; Timx 440 replace static fixture of local dataset

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context

Updating the load method to support partition prefixes enables PyArrow to only load data from partitions matching the prefix and avoid scanning the entire TIMDEXDataset.dataset. The added filtering method will be used in future read operations to retrieve relevant rows from TIMDEXDataset.dataset. (Taken from commit https://github.com/MITLibraries/timdex-dataset-api/pull/31/commits/41852bae34bf40b4d4abac57572d2d7384925c57 message).

More specifically, these changes will be required by `timdex-index-manager` (TIM) to filter rows in `TIMDEXDataset` for indexing records to and deleting records from OpenSearch. 

**I recommend thoroughly reading the docStrings written for [`TIMDEXDataset.load`](https://github.com/MITLibraries/timdex-dataset-api/pull/31/files#diff-fb40971da01d2dcbf97cdad92e2c3f6656d1eaa201505a2f9074e5d52e452142R87-R140) and [`TIMDEXDataset._get_filtered_dataset`](https://github.com/MITLibraries/timdex-dataset-api/pull/31/files#diff-fb40971da01d2dcbf97cdad92e2c3f6656d1eaa201505a2f9074e5d52e452142R176-R223) methods!** 📖 

This PR also contains additional work for [TIMX-440 Remove any previous partitioning strategy artifacts](https://mitlibraries.atlassian.net/browse/TIMX-440), which is addressed in the first commit https://github.com/MITLibraries/timdex-dataset-api/pull/31/commits/f2d3e96744efaee65fd9414d1a247e6576b00c44.

### How can a reviewer manually see the effects of these changes?

@ghukill Please proceed testing your in-progress read methods!
@ehanson8 Verifying that linters and unit tests are passing is sufficient. Worth noting are the following sections: 
   * [unit tests for `load` method](https://github.com/MITLibraries/timdex-dataset-api/pull/31/files#diff-26199baaf391361625bd6773b9df4ed2677df4680a9650d2731510adec4898eaR13-R122)
   * [unit tests for `filter` method](https://github.com/MITLibraries/timdex-dataset-api/pull/31/files#diff-26199baaf391361625bd6773b9df4ed2677df4680a9650d2731510adec4898eaR125-R209)
   * [unit tests for `get_partition_prefixes` method](https://github.com/MITLibraries/timdex-dataset-api/pull/31/files#diff-26199baaf391361625bd6773b9df4ed2677df4680a9650d2731510adec4898eaR212-R239)

### Includes new or updated dependencies?
NO

### Changes expectations for external applications?
YES - Any applications that use `timdex-dataset-api` and call `TIMDEXDataset.load` should be reviewed as the `load()` method no longer returns `self` and instead sets a `pyarrow.dataset.Dataset` on `self.dataset`.

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/TIMX-440
- https://mitlibraries.atlassian.net/browse/TIMX-425

### Developer
- [x] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [x] All related Jira tickets are linked in commit message(s)
- [ ] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed **or** provided examples verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: timdex_dataset_api/dataset.py [hunk 1]
```diff
@@ -41,31 +46,307 @@
...
+                    "'%Y-%m-%d' or a datetime.date."
+                )
+            filters_dict.update(
+                {
+                    "run_date": run_date_date,
+                    "year": run_date_date.strftime("%Y"),
+                    "month": run_date_date.strftime("%m"),
+                    "day": run_date_date.strftime("%d"),
+                }
+            )
```
Comment (ghukill) [inline timdex_dataset_api/dataset.py:267]: Noticed that this block is _very_ similar to this block from `get_partition_prefixes()`:

```python
if run_date:
    if isinstance(run_date, str):
        run_date_date = strict_date_parse(run_date)
    elif isinstance(run_date, date):
        run_date_date = run_date
    else:
        raise ValueError(
            "Provided 'run_date' value must be a string matching format "
            "'%Y-%m-%d' or a datetime.date."
        )
    return (
        f"year={run_date_date.strftime("%Y")}/"
        f"month={run_date_date.strftime("%m")}/"
        f"day={run_date_date.strftime("%d")}"
    )
```

Wonder if a private, helper method like `parse_date_filters()` could perform this work, and return a dictionary with all or some of `run_date, year, month, day` set in a dictionary?

**Review by ghukill — COMMENTED** [review]
This is looking real good.  Left a handful of comments about some possible tidying up, maybe consolidation of logic, but this operates as expected.

In a way, this is _the_ most opinionated part of the library.  Much of the read/write business logic is just relying on parquet libraries and patterns.  But here we are leaning heavily into our particular schema, and what kind of efficient querying patterns we want to support from the dataset.

Opted to submit a "Comment" review now, curious your thoughts on the comments so far.

File: timdex_dataset_api/dataset.py [hunk 2]
```diff
@@ -41,31 +46,307 @@
     "day",
 ]
 
+TIMDEX_DATASET_FILTER_COLUMNS = [
```
[the remark this answers is not in the document]
Comment (ghukill) [reply timdex_dataset_api/dataset.py:49]: **Clarification**: I think the explicit list is a good thing for now.  It was just _experimenting_ with that more dynamic building of the list that I realized the order was important.

**Comment by ghukill** [comment]
One last request @jonavellecuerdo - can you bump the version to `v0.4.0` as well?  I've [stubbed a ticket](https://mitlibraries.atlassian.net/browse/TIMX-449) to explore how we can streamline this version bumping (and it getting picked up by applications that install it), but just manual for now.

File: timdex_dataset_api/dataset.py [hunk 3]
```diff
@@ -41,31 +46,307 @@
...
+
+            - year, month, day
+            - year, month
+            - year
+
+            Any other combinations are insufficient to construct a valid partition prefix.
+
+        Returns a string of partition prefixes: "year=2024/month=12/day=01".
+        """
+        run_date_date: date | None = None
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply timdex_dataset_api/dataset.py:311]: You are right. Removed!

File: timdex_dataset_api/dataset.py [hunk 4]
```diff
@@ -41,31 +46,307 @@
...
+                cannot be parsed.
+
+        Returns:
+            ds.Dataset: Original pyarrow.dataset.Dataset (if no filters applied)
+                or new pyarrow.dataset.Dataset with applied filters.
+        """
+        if not self.dataset:
+            raise DatasetNotLoadedError
+
+        run_date_date: date | None = None
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply timdex_dataset_api/dataset.py:227]: Removed.

File: timdex_dataset_api/dataset.py [hunk 5]
```diff
@@ -41,31 +46,307 @@
...
+                ],
+                strict=False,
+            ),
+        )
+
+        # get filters for partition columns ('run_date' or 'run_date' components)
+        # update filters dict
+        if run_date is not None:
+            if isinstance(run_date, str):
+                run_date_date = strict_date_parse(run_date)
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply timdex_dataset_api/dataset.py:252]: Renamed! I wrestled with naming this variable. I wanted to reflect that it was A `datetime.date` object, but `run_date_dt_date` is more confusing, right?

File: timdex_dataset_api/dataset.py [hunk 6]
```diff
@@ -41,31 +46,307 @@
...
+            filters_dict.update(
+                {
+                    "run_date": run_date_date,
+                    "year": run_date_date.strftime("%Y"),
+                    "month": run_date_date.strftime("%m"),
+                    "day": run_date_date.strftime("%d"),
+                }
+            )
+        else:
+            filters_dict.update({"year": year, "month": month, "day": day})
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply timdex_dataset_api/dataset.py:269]: Good catch! I missed this while moving code around.

File: timdex_dataset_api/dataset.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply timdex_dataset_api/dataset.py:49]: Ah, that's a good catch. I updated to a tuple and added a comment:

```python
# order must match assignment in TIMDEXDataset._get_filtered_dataset
```

File: timdex_dataset_api/dataset.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ghukill) [reply timdex_dataset_api/dataset.py:252]: I think so.  Given that we have `run_date` everywhere (column name in dataset), and it's a string most of the time, I think it takes something quite explicit to indicate it's _not_ a string.  I think "object" or "obj" is pretty universal for, "something complex, not a string or integer."

File: timdex_dataset_api/dataset.py (hunk 1 above)
Comment (jonavellecuerdo) [reply timdex_dataset_api/dataset.py]: A[ new private method `parse_date_filters()`](https://github.com/MITLibraries/timdex-dataset-api/pull/31/commits/0cc15c360be5f984b0461beb5826bdd158df46da#diff-fb40971da01d2dcbf97cdad92e2c3f6656d1eaa201505a2f9074e5d52e452142R327-R347) has been added.

**Review by ghukill — CHANGES_REQUESTED** [review]
Overall, the changes are precisely what we discussed, and well implemented.

However, during some testing with the 5-year simulated dataset in Dev1 S3 (`s3://timdex-extract-dev-222053980223/dataset-five-year-simulation/`), an unexpected edge case cropped up.

The following is an attempt at a summary of discussions I had with @jonavellecuerdo when tracking this down.

In short: when a prefix is constructed for the dataset with `year`, `month`, and `day` (or a combination of them), we get `None` values for any partition columns used in the prefix.  This isn't inherently a bug, but it is then problematic with the "post-load" filtering we apply which _again_ attempts to filter on those partition columns.  It was believed that would be somewhat inconsequential, but it was not.

Here is an example:

```python
from timdex_dataset_api import TIMDEXDataset

DATASET_LOCATION='s3://timdex-extract-dev-222053980223/dataset-five-year-simulation/'
td = TIMDEXDataset(DATASET_LOCATION)

td.load(year='2025', month='06', day='02')
td.row_count # returns 0, which is unexpected
td.dataset.count_rows() # same for direct call
```

It's subtle.  The "pre-load" work of adding a prefix to the dataset works, and we get _only_ the parquet files that match those YMD values.  But then the "post-load" filtering effectively filters out all rows, because those columns have `None` values.  It can loosely be understood as by providing more of the parquet filename path in the dataset location, we remove the ability for parquet libraries to **extract** partition values from the filenames themselves.  It's subtle, because not all parquet reading libraries handle this quite the same.  I believe DuckDB would still extract the partition values even with a "deeper" root path, AWS Glue as well.  But given our heavy usage of `pyarrow`, we should design towards that.

While there are technically workarounds to this, it also hints at an anti-pattern this dataset prefixing as part of the "pre-load" work introduces.  To some degree, we are short-circuiting the strength of parquet datasets and how libraries like `pyarrow` or `DuckDB` interact with them.  They excel when they are presented with the "full" dataset, and then intelligently utilize partitions in the filenames to prune data before reading.

Taking all this, and some additional nuances not well represented here but touched in a discussion, the following is proposed:

1. **remove** all "pre-load" dataset location filtering
2. **keep** all columns in the schema and columns to filter by, including `year`, `month`, and `day`

The effects of this proposal:
- initial load time of the dataset is slightly slower, as all files in the dataset are discovered (this was later understood to not be that bad, where 13k files for 5-year simulation load in 1.68s)
- any operations that `pyarrow` or `DuckDB` perform will have full access to the full dataset, relying on their partition and column filtering

Thanks to @jonavellecuerdo for the flexibility in this area and PR as we learn some of the nuances of working with a parquet dataset!

**Comment by ehanson8** [comment]
> Overall, the changes are precisely what we discussed, and well implemented.
> 
> However, during some testing with the 5-year simulated dataset in Dev1 S3 (`s3://timdex-extract-dev-222053980223/dataset-five-year-simulation/`), an unexpected edge case cropped up.
> 
> The following is an attempt at a summary of discussions I had with @jonavellecuerdo when tracking this down.
> 
> In short: when a prefix is constructed for the dataset with `year`, `month`, and `day` (or a combination of them), we get `None` values for any partition columns used in the prefix. This isn't inherently a bug, but it is then problematic with the "post-load" filtering we apply which _again_ attempts to filter on those partition columns. It was believed that would be somewhat inconsequential, but it was not.
> 
> Here is an example:
> 
> ```python
> from timdex_dataset_api import TIMDEXDataset
> 
> DATASET_LOCATION='s3://timdex-extract-dev-222053980223/dataset-five-year-simulation/'
> td = TIMDEXDataset(DATASET_LOCATION)
> 
> td.load(year='2025', month='06', day='02')
> td.row_count # returns 0, which is unexpected
> td.dataset.count_rows() # same for direct call
> ```
> 
> It's subtle. The "pre-load" work of adding a prefix to the dataset works, and we get _only_ the parquet files that match those YMD values. But then the "post-load" filtering effectively filters out all rows, because those columns have `None` values. It can loosely be understood as by providing more of the parquet filename path in the dataset location, we remove the ability for parquet libraries to **extract** partition values from the filenames themselves. It's subtle, because not all parquet reading libraries handle this quite the same. I believe DuckDB would still extract the partition values even with a "deeper" root path, AWS Glue as well. But given our heavy usage of `pyarrow`, we should design towards that.
> 
> While there are technically workarounds to this, it also hints at an anti-pattern this dataset prefixing as part of the "pre-load" work introduces. To some degree, we are short-circuiting the strength of parquet datasets and how libraries like `pyarrow` or `DuckDB` interact with them. They excel when they are presented with the "full" dataset, and then intelligently utilize partitions in the filenames to prune data before reading.
> 
> Taking all this, and some additional nuances not well represented here but touched in a discussion, the following is proposed:
> 
> 1. **remove** all "pre-load" dataset location filtering
> 2. **keep** all columns in the schema and columns to filter by, including `year`, `month`, and `day`
> 
> The effects of this proposal:
> 
> * initial load time of the dataset is slightly slower, as all files in the dataset are discovered (this was later understood to not be that bad, where 13k files for 5-year simulation load in 1.68s)
> * any operations that `pyarrow` or `DuckDB` perform will have full access to the full dataset, relying on their partition and column filtering
> 
> Thanks to @jonavellecuerdo for the flexibility in this area and PR as we learn some of the nuances of working with a parquet dataset!

Very good write up, I appreciate the context!

**Comment by ghukill** [comment]
Another small request @jonavellecuerdo.  Unrelated to your PR, but I think would be helpful to sneak in, as it was discovered while testing installation of this library.

Can you add `attrs` to the dependencies used for building here?
https://github.com/MITLibraries/timdex-dataset-api/blob/TIMX-425-update-load-dataset-and-apply-filtering/pyproject.toml#L24-L29

**Review by ghukill — APPROVED** [review]
Approved!

All work as expected:

```python
import datetime

from timdex_dataset_api import TIMDEXDataset

DATASET_LOCATION='s3://timdex-extract-dev-222053980223/dataset-five-year-simulation/'
td = TIMDEXDataset(DATASET_LOCATION)

# without any filters
td.load()

# with YMD
td.load(year="2025",month="06",day="02")

# with run-date string
td.load(run_date="2025-06-02")

# with run-date datetime obj
td.load(run_date=datetime.date(2025,6,2))
```

Moreover, to your point in slack, there is a distinct simplification of code and logic.  Again, thanks for the patience while wading through this.  This was the time to get more familiar and comfortable with the idiosyncracies of a parquet dataset, and this library is feeling sound for it.

Next step: read methods!  Which will draw directly from these additions, and finally close the loop of library responsibilities.

File: timdex_dataset_api/dataset.py [hunk 7]
```diff
@@ -41,31 +45,257 @@
...
+        expressions = []
+        for field, value in filters_dict.items():
+            if value:
+                expressions.append(pc.equal(pc.field(field), value))
+
+        # if filter expressions not found, return original dataset
+        if not expressions:
+            return self.dataset
+
+        # combine filter expressions as a single predicate
```
Comment (ehanson8) [inline timdex_dataset_api/dataset.py:258]: These comments are very helpful in making this method's functionality clear!

**Review by ehanson8 — APPROVED** [review]
Approved, great work!

**Pull request merged**